inject.js

```diff
@@ -369,7 +398,11 @@
           }
 
           // Ignore keydown event if typing in a page without vsc
-          if (!document.querySelector(".vsc-controller")) {
+          if (document.querySelector('apple-tv-plus-player')) {
```

Comment:
good question. even if you find a shadowRoot in an element, you have to recursively search inside to find shadowRoots in all of the shadowRoot elements. On a page that has lots of nested shadowRoots, this might result in a performance hit.  
https://jsperf.com/treewalker-vs-dom-traversal/25
The above test tests the performance of a shallow (not shadow) DOM, and nodeIterator is faster than treeWalker.
A previous test https://jsperf.com/treewalker-vs-dom-traversal/9 tested a treeWalker versus a for loop over an array (specifically, a NodeList collection) of elements in document.body.getElementsByTagName("*"), and the treeWalker was 7160x faster.
https://jsperf.com/recursion-vs-iteration-852/31
The above test tests the performance of deeply nested div's in a DOM, and the treeWalker is 14% faster than nodeIterator but a simple recursive call is the fastest.
```js
var nodes=[]; 
function walkTheDOM(node) {
     nodes.push(node);
     node = node.firstChild;
     while (node) {
         walkTheDOM(node);
         node = node.nextSibling;
     } 
} 
walkTheDOM(parentElement);
```
on my machine, walkTheDOM is 20% faster than treeWalker, but this test is bad because it doesn't use the same number of divs between the tests. It uses Math.random(). I wouldn't trust this test.

Reply:
I appreciate seeing all of the detailed performance notes, I'm definitely learning just being a bystander here. In my opinion, to address @igrigorik's question I don't think performance should be a reason for making the implementation apple exclusive. Perhaps if it truly presents itself as a performance concern we could add a setting to toggle scanning shadowroots (keep in mind, this only impacts sites which actually have shadowroots).

That said, in theory the addon isn't necessarily doing more work. If I design a website with 1000 lines in the DOM, then refactor it to place 500 of those including video into a shadowroot, VSC would then become twice as fast and 100% more useless. I realize that may not work out to be exactly twice as fast, but a close approximation

Reply:
@jedawson thanks for the detailed analysis. @ChadBailey's reply captures where I was coming from..

Hypothetically, let's assume Apple TV folks did not choose to use Web Components (but I'm glad they did), we'd have the usual nested div-soup to deal with: ~same nesting minus the encapsulation. Yes, getElement* would be able to do a faster lookup at first but the mutation observer would still recursively traverse the mutation record. Which is to say, if they did not use Web Components we'd just assume the cost and go forward as usual. At the same time, special casing Apple TV here — my bet — doesn't help in the long term, because I would expect more and more sites to adopt Web Components and encapsulate their vide player logic. As a result, my recommendation is to make this agnostic: it makes sense for us to traverse all shadowRoots and look for video+audio.

WDYT, does that make sense?

Reply:
I agree with you @igrigorik . Not only does it ensure VSC has better support for all sites, but it also ensures we don't have to redesign every time apple TV decides they want to make a change that breaks VSC.

Reply:
What if you're on facebook, which has like 12000 elements, it wouldn't make sense to loop through each element to see if there is a shadow root. is there a way to see if there is a video playing on the screen? Is there something like IntersectionObserver? with IntersectionObserver, you have to know the video tag beforehand, so it might not work without knowing the video tag.

Reply:
https://github.com/whatwg/dom/issues/665

Reply:
@jedawson would it be a lot of effort to update the code in a platform agnostic way so we can test it? I have 2 very underpowered machines (an Intel Atom and Chromebook) I could test on sites with lots of deep nesting like Facebook.

Reply:
ok. sure!
I'm also trying to figure out something strange. The ShadowRoot is showing up as 
```js 
nodeName: "#document-fragment"
nodeType: 11
```
which is strange because on MDN, it says that `NodeFilter.SHOW_DOCUMENT_FRAGMENT` is `1024`

Reply:
I'm going to create a function that will get the children and do getElementByTagName on each and concat them together and look for shadowroot and recursively do it again.

Reply:
```js
function queryDeep(element){
	let result = []
	if (element.shadowRoot) {
		result = result.concat(queryDeep(element.shadowRoot))
    }
	for (let child of element.children) {
		if (child.shadowRoot) {
			for (let grandChild of child.shadowRoot.children){
				result = result.concat(queryDeep(grandChild))
			}
        }
        if (child.tagName == 'VIDEO') {
            result = result.concat(child)
        } 
        if (child.hasChildNodes()) {
            for (let grandChild of child.children) {
                result = result.concat(queryDeep(grandChild))
            }
        }
	}
	return result
}
queryDeep(document.body)
```

Reply:
this doesn't work for any standard website that doesn't have any video tags in the light dom.
also, I am changing this to use while loops instead of for loops

Reply:
this is a good explanation https://chromium.googlesource.com/chromium/src/third_party/+/master/blink/renderer/core/dom/README.md

Reply:
Thanks for the continued work on this @jedawson! Let me know when it's ready to be tested and I'll take it for a spin.

In reference to the performance, I had the idea that if this does turn out to be a computationally expensive task perhaps we could offload it to wasm? I've never used it so I don't know if that's even possible - just a thought. I'd be glad to take a closer look if it proves necessary

Reply:
I'm not sure how bad the performance hit would be since I don't know how these things are implemented, but I think Chad/Ilya's point is that even if we have 15 mutation observers it could be equivalent/better than if the same sized DOM and DOM changes were all monitored by 1 observer. I'd imagine there is some crossover where this is no longer true and the overhead of 100+ observers is a problem.

Reply:
@igrigorik how do CSS renderers for browsers handle this?

It would be super nice if we could just override how chrome implements the video tag....but I assume that would create many issues for everyone else.

Reply:
> I'm not sure how bad the performance hit would be since I don't know how these things are implemented, but I think Chad/Ilya's point is that even if we have 15 mutation observers it could be equivalent/better than if the same sized DOM and DOM changes were all monitored by 1 observer. I'd imagine there is some crossover where this is no longer true and the overhead of 100+ observers is a problem.

here is kind of answer to your question. 
https://stackoverflow.com/questions/31659567/performance-of-mutationobserver-to-detect-nodes-in-entire-dom

Reply:
i was able to get the div.vsc-controller to show the controls when moused over by getting it out of the shadow root and putting it as the first child of the shadow root's parent. I will need to change some code so that the buttons work.

Reply:
> how do CSS renderers for browsers handle this?

this doesn't involve CSS. do you mean the javascript engine V8?

Reply:
Naww, I don't know really know what I meant. Something about browser renderers having to keep track of all dom changes to see if css selectors are true.

Reply:
I think you might have read something about querySelector and getElement*. The querySelector returns a static NodeList, getElements* returns a live HTMLCollection.